### DISCUSSION TITLE
RFC - add rawBody to NextApiRequest

### DISCUSSION THREAD

**Discussion by mikesol** [vercel/next.js#13405] [Help] [answered]
# Feature request

As per @timneutkens request, this is an RFC to add `rawBody` to NextApiRequest.

## Is your feature request related to a problem? Please describe.

Currently, there is no way to use both a body parser _and_ recuperate the raw body of a request. This makes things like [securing github webhooks](https://developer.github.com/webhooks/securing/), which rely on calculating a hash based on a raw body, difficult in NextJS.

## Describe the solution you'd like

The solution I proposed in the PR above is to add a raw body field to NextApiRequest.

## Describe alternatives you've considered

The alternative we're currently using is deploying a separate microservice to handle our github webhook. It could be that this feature is outside the scope of nextjs, in which case the separate microservice makes sense.

**Comment by Timer** [comment]
Would [disabling the `bodyParser`](https://nextjs.org/docs/api-routes/api-middlewares#custom-config) be an appropriate action?

This allows you to read the body however you'd like.

**Comment by mikesol** [comment]
Hey! This definitely helps - we were hoping to use the body parser as the body is json, but what we can do as a temporary measure is disable the body parser and then parse the JSON ourselves. Thanks!

**Comment by eknowles** [comment]
I would add to this, I had the same issue with stripe webhooks.
My solution was to reverse the process as described here https://github.com/stripe/stripe-node#testing-webhook-signing
`JSON.stringify(payload, null, 2)`.

**Reply by colllin** [reply]
I believed it.  I tried it.  It didn't work.  This makes sense — it works in the test because we selected which string to sign (e.g. ~`JSON.stringify(req.body, null, 2)`~), but doesn't work when we don't know exactly which json string was signed. Apparently, based on my testing, Stripe does not use ~`JSON.stringify(req.body, null, 2)`~ in their API.  That also makes sense, because it would waste bandwidth with unnecessary characters.

**Reply by rahmanqureshi** [reply]
yeah it also doesn't work for me.

**Comment by saiichihashimoto** [comment]
The stringify solution did not work for me. Also, after disabling the bodyParser, `req.body` is undefined, how would one parse the body?

**Reply by davecardwell** [reply]
See https://github.com/vercel/next.js/blob/86160a5190c50ea315c7ba91d77dfb51c42bc65f/test/integration/api-support/pages/api/no-parsing.js for an example.

**Reply by saiichihashimoto** [reply]
It's odd that the default way, that must have the rawBody in order to make the body, wouldn't expose it. What happens if I want both the body and rawBody? I'd be forced to reinvent what next.js already does, which isn't ideal.

**Reply by davecardwell** [reply]
I guess by allowing the default behavior to be overwritten it leaves the ability for things like implementing a streaming parser for large bodies, writing the file to disk without reading it all into memory for file uploads, etc?

**Reply by a14m** [reply]
Thanks @davecardwell for pointing out this https://github.com/vercel/next.js/blob/86160a5190c50ea315c7ba91d77dfb51c42bc65f/test/integration/api-support/pages/api/no-parsing.js  link which helped fixing the issue I had with validating the webhooks signatures

```js
const webhookPayloadParser = (req: NextApiRequest) =>
  new Promise((resolve) => {
    let data = "";
    req.on("data", (chunk) => {
      data += chunk;
    });
    req.on("end", () => {
      resolve(Buffer.from(data).toString());
    });
  });
```
and then it can be used `const data = await webhookPayloadParser(req);` in different pages

**Reply by alicescfernandes** [reply]
Hi. I stumbled upon the same issue and had to come up with a way to validate Github webhooks. I've taken the suggestion of @davecardwell and expanded so it would verify the signature of the github webhooks. You can take a look at the final piece of code on https://github.com/alicescfernandes/mapa-vacinacao-c19/blob/d693819d952df5519cef93d4ada531677172711e/pages/api/hooks.js#L22, but please note the environment variables that are used across this implementation

To achieve this, first i tried to use this [middleware](https://github.com/compwright/x-hub-signature), and did managed to pass through the raw body data, but i kept getting the `req.header is not a function` error, so i just ended up using this [gist](https://gist.github.com/stigok/57d075c1cf2a609cb758898c0b202428) along the code that @davecardwell put above. 
It works like a charm

**Reply by devopsangel** [reply]
> Hi. I stumbled upon the same issue and had to come up with a way to validate Github webhooks. I've taken the suggestion of @davecardwell and expanded so it would verify the signature of the github webhooks. You can take a look at the final piece of code on https://github.com/alicescfernandes/mapa-vacinacao-c19/blob/d693819d952df5519cef93d4ada531677172711e/pages/api/hooks.js#L22, but please note the environment variables that are used across this implementation
> 
> To achieve this, first i tried to use this [middleware](https://github.com/compwright/x-hub-signature), and did managed to pass through the raw body data, but i kept getting the `req.header is not a function` error, so i just ended up using this [gist](https://gist.github.com/stigok/57d075c1cf2a609cb758898c0b202428) along the code that @davecardwell put above.
> It works like a charm

@alicescfernandes did u get both `body` and `rawBody` on req?  I need both ... if I disable  `bodyParser: false` it works however I am loosing `req.body`.  Dunno what to do ...

**Reply by alicescfernandes** [reply]
@ devopsangel not exactly req.body but what i did was parse the `rawBody`
```js
let json_string = decodeURIComponent(req.rawBody).split('payload=')[1];
let json = JSON.parse(json_string);
```

What you could do was something like
```js
export default async function handler(req, res) {
	const data = await webhookPayloadParser(req);
	req.rawBody = data;
	
	let json_string = decodeURIComponent(req.rawBody).split('payload=')[1];
        let json = JSON.parse(json_string);
        req.body = json
        
        //run your code with the modified req

}
```
Then you would have the `body` as usual. This solution still needs some cleanup though, but it might work for you

**Reply by devopsangel** [reply]
@alicescfernandes thank you!

**Reply by Noitidart** [reply]
We should probably add how to get rawBody into the docs. I expected it to just be on `req.body`.

**Comment by jhookom** [comment]
Are there any future plans/alternate ways of maybe defining interceptors at the config for handling hashes/security on a per API call?  I'm sure many are writing wrapper methods like we are to handle this, while requiring the developer to also disable body parsing each time to work properly.

https://github.com/jhookom/nextjs-oc-api/blob/main/pages/api/checkout/%5B...event%5D.ts#L6

https://github.com/jhookom/nextjs-oc-api/blob/main/lib/oc-catalyst-next.ts#L247

`

**Comment by robert113289** [comment]
disabling the parser means that reg.body will be undefined 
we fixed this by adding in middleware like below. the line with .apply(text({ type: '*/*' })) is adding in the middleware that will set req.body to a string when the request comes in. 

```
async function bootstrap() {
  const app = await NestFactory.create(AppModule, { bodyParser: false });
  await app.listen(3000);
}
```

```
export class AppModule {
  configure(consumer: MiddlewareConsumer): void {
    consumer
      .apply(text({ type: '*/*' }))
      .forRoutes('/');
  }
}
```

**Reply by mcgraphix** [reply]
where is that `text()` function coming from?

**Reply by sacru2red** [reply]
this is nextjs issue page. your code is nestjs.

**Comment by wicka-aus** [comment]
Hi 

pulling my hair out here trying to make this Stripe Webhook work.

Tried all of the above!

Secret is fine.
Signatures exist.

Cant pass JSON req into raw as I am still getting error -
"Webhook signature verification failed. No signatures found matching the expected signature for payload. Are you passing the raw request body you received from Stripe?"

What am I missing here??

TIA

```
import { buffer } from 'micro';
const stripe = require('stripe')(process.env.STRIPE_SECRET_KEY);

export default async function handler(req, res) {

    console.log("Payment intent")
    console.log(process.env.STRIPE_WEBHOOK_ENDPOINT_SECRET)

    const buf = await buffer(req);

    console.log(buf)

    let event

    if (process.env.STRIPE_WEBHOOK_ENDPOINT_SECRET) {
        // Get the signature sent by Stripe
        const signature = req.headers['stripe-signature'];
        console.log(signature)
        try {
          event = stripe.webhooks.constructEvent(
            buf,
            signature,
            process.env.STRIPE_WEBHOOK_ENDPOINT_SECRET
          );
        } catch (err) {
          console.log(`⚠️  Webhook signature verification failed.`, err.message);
          return res.status(400);
        }
      }

    console.log(event.type)

    // Handle the event
    switch (event.type) {
        case 'payment_intent.succeeded':
            console.log("yes")
            res.status(200);   
            break;

        default:
            console.log(`Unhandled event type ${event.type}`);
    }



}


export const config = {
    api: {
      bodyParser: false,
    },
  }
  
```

**Reply by dalmo3** [reply]
Check out https://github.com/vercel/next.js/discussions/13405#discussioncomment-1668455 below if you're still stuck on this.

**Comment by dalmo3** [comment]
Just stumbled upon this issue and got it to work with the `raw-body` package.

In my use case, verifying stripe webhooks, I don't need to parse the body. So I'm not sure this solves all use cases but since many others on this thread mentioned stripe I decided to share it.

```typescript
import getRawBody from 'raw-body';

export default async function handler(req, res) {
    const rawBody = await getRawBody(req);

    const endpointSecret = ...;
    const signature = req.headers['stripe-signature'] ?? '';

    const event = stripe.webhooks.constructEvent(
      rawBody,
      signature,
      endpointSecret
    );
}  

export const config = {
  api: {
    bodyParser: false,
  },
};

```

**Reply by danielcamargo** [reply]
Look no further :D

**Reply by Koyamie** [reply]
Thank you.

**Reply by ValentineCodes** [reply]
Thanks. Finally I get a different error message. We keep googling

Have you encountered "Webhook Error: Timestamp outside the tolerance zone"?
If so, How'd you deal with it?

**Reply by italodeandra** [reply]
No need to use external library for it. Here is the `getRawBody` function:

```ts
async function getRawBody(readable: Readable): Promise<Buffer> {
  const chunks = [];
  for await (const chunk of readable) {
    chunks.push(typeof chunk === "string" ? Buffer.from(chunk) : chunk);
  }
  return Buffer.concat(chunks);
}
```

**Comment by ondrabus** [comment]
If none of the above worked for you, you may try a solution from [CodeDaily](https://codedaily.io/tutorials/Stripe-Webhook-Verification-with-NextJS)

Disable the `bodyParser`
```
export const config = {
  api: {
    bodyParser: false,
  },
};
```

Use [micro by Vercel](https://github.com/vercel/micro) to get the body:
```
npm i micro
```

```
import { buffer } from "micro";
...
const body = (await buffer(req)).toString();
```

(I also need the parsed body, so after checking the signature, I do `const data = JSON.parse(body)`)

**Comment by techygrrrl** [comment]
I also had this problem trying to parse the Stripe webhook in Next.js. 

I resolved it by using body-parser's `raw()` as middleware—`bodyParser.raw({type: 'application/json'})`.

First I used a generic middleware function to help with applying middleware to individual routes:

```ts
// ./utils/middleware.utils.ts
export function runMiddleware(req: NextApiRequest, res: NextApiResponse, fn: Function) {
  return new Promise((resolve, reject) => {
    fn(req, res, (result: Object) => {
      if (result instanceof Error) {
        return reject(result)
      }

      return resolve(result)
    })
  })
}
```

I disabled Next.js's implementation of body-parser and implemented `bodyParser.raw()`:

```ts
// ./pages/api/webhooks.ts

// this disables bodyParser
export const config = {
  api: {
    bodyParser: false,
  }
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<ResponseData>
) {
  // use bodyParser.raw() as middleware instead
  await runMiddleware(req, res, bodyParser.raw({ type: 'application/json' }))

  try {
    const signature = req.headers['stripe-signature']

    const event = stripe.webhooks.constructEvent(req.body, signature, process.env.STRIPE_WEBHOOKS_SECRET)

    res.status(200).json({})

    // TODO: Handle event
  } catch (e) {
    return res.status(400).json({ error: 'Webhook error' })
  }    
}
```

Using `bodyParser.raw({type: 'application/json'})` aligns with the Express.js example in Stripe's docs: https://stripe.com/docs/payments/checkout/fulfill-orders#verify-events-came-from-stripe

```js
// Stripe example from their docs
app.post('/webhook', bodyParser.raw({type: 'application/json'}), (request, response) => {
  // 
})
```

**Comment by leerob** [comment]
https://github.com/vercel/next.js/discussions/27702#discussioncomment-2929911

**Comment by tordsta** [comment]
Found this thread while working on stripe webhook. I'll add the solution i used, I found a simple way to solve this with the raw-body package. Don't know if is better than leerob's suggestion above, I suggest checking it out as well. 

```
import { NextApiRequest, NextApiResponse } from "next";
import getRawBody from "raw-body";

export const config = {
  api: {
    bodyParser: false,
  },
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<string>
) {
  const rawBody = await getRawBody(req);
  const sig = req.headers["stripe-signature"];
  const endpointSecret = process.env.ASDF

  let event;

  try {
    event = stripeServer.webhooks.constructEvent(rawBody, sig, endpointSecret);
  } catch (err) {
    console.log(err);
    res.status(400).send(`Webhook Error: ${err}`);
    return;
  }
  // continue your code here

```

**Comment by ben519** [comment, marked as answer]
This worked for me, for handling a stripe webhook in a route handler.

```js
export async function POST(request) {
  const rawBody = await request.text()
  // ...
}
```

No special configuration needed. (Next `13.5` app dir)

**Reply by ashxjs** [reply]
updated from `next@13.4.19` to `next@13.5` successfully worked ! thank's @ben519

**Reply by abdulbari149** [reply]
This isn't working for me

**Reply by siamahnaf** [reply]
not working!

**Comment by siamahnaf** [comment]
```
const WEBHOOK_SECRET = process.env.STRIPE_WEBHOOK_SECRET!

export const POST = async (req: NextRequest) => {
    const rawBody = await req.text();
    const sig = req.headers.get("stripe-signature")!;

    let event: Stripe.Event;

    try {
        event = Stripe.webhooks.constructEvent(rawBody, sig, WEBHOOK_SECRET);
    } catch (e) {
        return new Response("Webhook Error", { status: 400 });
    }

    return Response.json({ success: true });
}
```

It worked only on local listener. not on live link.


Error is-

Error: No signatures found matching the expected signature for payload. Are you passing the raw request body you received from Stripe? 
 If a webhook request is being forwarded by a third-party tool, ensure that the exact request body, including JSON formatting and new line style, is preserved.

Learn more about webhook signing and explore webhook integration examples for various frameworks at https://docs.stripe.com/webhooks/signature


It is nextjs 15.3.3 and app router.
